In-form path choice

cvalarida opened the pull request:
## Description
The veteran can now choose which path to go down.

The form should now reflect @goldenmeanie's diagram: 
![image](https://user-images.githubusercontent.com/12970166/51949528-b56dda80-23e1-11e9-85f3-6d0649513d67.png)


## Testing done
Unit and manual testing

## Screenshots
![image](https://user-images.githubusercontent.com/12970166/51948803-d5e86580-23de-11e9-8137-a61470225157.png)

@peggygannon What should this page say? I copied the checkbox content from the wizard.
Also, what should the disability orientation page say? (I originally asked the question in the [linked ticket](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/16466#issuecomment-458315997).)

## Acceptance criteria
- [x] The veteran can choose which path of the form to go down

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Resolves https://github.com/department-of-veterans-affairs/vets.gov-team/issues/16465

Commit:
Fix unrelated typo

Commit:
Add new page

Commit:
Fix propTypes warning

Commit:
Change to checkboxes

Commit:
Add path choice page and depends callbacks

peggygannon reviewed (changes requested):
@cvalarida Hi, Chris, please see my content suggestions below
 I added a checkbox for filing the first disability claim. This should be one of the options, right? 
 
Please tell us what type of disability claim you're filing. (Select all that apply.)

{checkbox} This is the first disability claim I'm filing. 
{checkbox} I have a new condition to add to my rated claim. 
{checkbox} One or more of my rated conditions has gotten worse.

peggygannon commented:
@cvalarida Ok, thanks. Yes, let's remove "rated." I think that would work I also tweaked the checkbox answers a bit to better answer the question.
See below and let me know if you think this would work:  
 
Please tell us what type of disability claim you're filing. (Select all that apply.)
{checkbox} I'm filing a claim for a new condition. 
{checkbox} I'm filing a claim because one or more of my rated conditions has gotten worse.

goldenmeanie commented:
@peggygannon is there a way you are comfortable with, to remove the duplication (I'm filing a claim) in the options?

@cvalarida:
1. Does `type-of-claim` make sense to you and @peggygannon?
2. Will look at that
3. I think this gets at the required question. If you don't answer, then you get the increase and new flows?

cvalarida commented:
@peggygannon That looks great!

@goldenmeanie 
1) `type-of-claim` looks great to me
2) :+1: 
3) Right. I'll verify that's the way it works now, but assuming that's the way we _want_ it to work, how does that sound to you?

Commit:
Url change

Commit:
Refactor tests

Commit:
Hide orientation screen

peggygannon commented:
@cvalarida @goldenmeanie here's the content update to the wizard question/answers. Per our style on the static site, we bold the repeated set up to the answers. 
Let me know if you think this will work. We might have to tweak slightly when we add BDD and appeals checkboxes. 

Please tell us what type of disability claim you're filing. (Select all that apply.)

**I'm filing a claim:** 
{checkbox} For a new condition
{checkbox} Because one or more of my rated conditions has gotten worse

Commit:
Remove `notIncreaseOnly` and `notNewOnly`

Commit:
Formatting

Commit:
Rename to claimType

Commit:
Claim type content change

cvalarida commented:
This PR now has the following as the second page in the form (right after veteran information):
![image](https://user-images.githubusercontent.com/12970166/51939496-741b0200-23c4-11e9-90a7-bd6ea4b4188d.png)

peggygannon commented:
@cvalarida the screenshot is missing the answer setup (below in bold). it should appear on its own line.

**I'm filing a claim:**
{checkbox} For a new condition
{checkbox} Because one or more of my rated conditions has gotten worse

Commit:
Missed a spot

cvalarida commented:
Oh dang it. I thought I did that part too. Good catch.
![image](https://user-images.githubusercontent.com/12970166/51940273-66ff1280-23c6-11e9-90b0-b42426f8e03a.png)

cvalarida commented:
Ah, blast. I did miss that part. How's this look?
![image](https://user-images.githubusercontent.com/12970166/51940529-2653c900-23c7-11e9-9ee2-59147e56b900.png)

Commit:
Still missed a spot

Commit:
Use .jsx extension

peggygannon commented:
@cvalarida Looks good! thank you!

peggygannon commented:
@cvalarida hi, Chris, I provided updated copy in this issue:
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/16466

Can you take a look and let me know if you have any questions? I also would like @goldenmeanie thoughts on the new disability orientation screen copy.

Commit:
Remove some content on orientation

Commit:
Move page in index

Commit:
Require a claim type

Commit:
Add comment

Commit:
Add migration

cvalarida commented:
After I added the validation, it changed up the formatting a bit, so for posterity, here it is:
![image](https://user-images.githubusercontent.com/12970166/51948803-d5e86580-23de-11e9-8137-a61470225157.png)

With the validation message:
![image](https://user-images.githubusercontent.com/12970166/51948815-e0a2fa80-23de-11e9-9ee4-79725fe34340.png)

Commit:
Update orientation content

cvalarida commented:
@peggygannon I updated the orientation screen per the ticket you linked: 
![image](https://user-images.githubusercontent.com/12970166/51949057-ddf4d500-23df-11e9-966e-54ada12b2254.png)

I also updated the rated disabilities screen like you mentioned:
![image](https://user-images.githubusercontent.com/12970166/51949127-3330e680-23e0-11e9-9b7d-a14fecad97d6.png)

Commit:
Update ratedDisabilities content

cvalarida commented on src/applications/disability-benefits/all-claims/components/IntroductionPage.jsx:
```diff
@@ -180,7 +180,7 @@ function mapDispatchToProps(dispatch) {
 }
 
-IntroductionPage.PropTypes = {
+IntroductionPage.propTypes = {
```
Removed a React warning

cvalarida commented on src/applications/disability-benefits/all-claims/config/form.js:
```diff
@@ -108,10 +116,9 @@ const formConfig = {
-  version: 1,
-  migrations: [],
+  formId: '21-526EZ',
+  version: migrations.length,
```
The form version will always be the length of the migrations until we remove some.

The reason `version` is separate from `migrations` is because we were thinking we'd remove them later. The current implementation of SiP migrations assumes they'll never get removed, though, so this is likely to stay like this indefinitely. If we do later remove old migrations to reduce the bundle size, we'll likely replace them with `savedForm => savedForm`. :man_shrugging:

markgreenburg replied:
Huh. TIL.

cvalarida commented on src/platform/forms/save-in-progress/actions.js:
```diff
@@ -142,7 +142,7 @@ export function migrateFormData(savedData, migrations) {
   let savedVersion = savedData.metadata.version;
   while (typeof migrations[savedVersion] === 'function') {
     savedDataCopy = migrations[savedVersion](savedDataCopy);
-    savedVersion++; // eslint-disable-line no-param-reassign
```
Just a bit of cleanup while I was looking at it; this is no longer a param, so no need to ignore it. :man_shrugging:

cvalarida commented on src/platform/testing/e2e/form-tester/form-filler.js:
```diff
@@ -34,7 +34,7 @@ const getElementSelector = (field, fieldData) => {
       typeof fieldData === 'boolean' ? (fieldData ? 'Y' : 'N') : fieldData
     }"]`,
     // Date has two or three elements, but should always have a year
-    // Return a valid selector so it doesn't get skippeG
```
More cleanup. Has nothing whatsoever to do with this PR, but I didn't want to just leave it.

bshyong commented on src/applications/disability-benefits/all-claims/config/form.js:
```diff
@@ -238,7 +258,7 @@ const formConfig = {
         newDisabilityFollowUp: {
           title: formData => capitalizeEachWord(formData.condition),
-          depends: form => form['view:newDisabilities'] === true,
+          depends: hasNewDisabilities,
```
👍

peggygannon commented:
Hi, @cvalarida Alex and I synced on this yesterday we have some additional copy tweaks: 

1. Updated copy for in-form wizard: 
Please tell us what type of disability claim you're filing. (Select all that apply.)
**I'm filing a claim for:**
{checkbox} A new condition
{checkbox} One or more of my rated conditions that have gotten worse

2. We want to use number rather than bullets for the choices below: 

**Disability orientation screen** 
One the next few screens, we'll ask you about the disabilities you're claiming in two parts: 
1. Rated service-connected disabilities. We'll ask you to tell us which rated disability you're claiming for increased compensation. 
2. New service-connected disabilities or conditions. 

Once we have your list of service-connected disabilities, we'll ask you more specific questions about each of them.

Commit:
Update content on the claimType page

Commit:
Use an ordered list

cvalarida commented:
![image](https://user-images.githubusercontent.com/12970166/52379566-ac06f280-2a1f-11e9-8dfa-c1fe89c499d3.png)
![image](https://user-images.githubusercontent.com/12970166/52379637-dc4e9100-2a1f-11e9-8c24-8cdf29748289.png)

peggygannon commented:
@cvalarida sorry, i didn't get back to you earlier. 

for the 2nd screen, we don't want **Your** anymore. I also have a slight tweaks the language in the steps below. (sorry!) 
It should look like this: 

On the next few screens, we’ll ask you about the disabilities you’re claiming, in two parts.

1. Rated service-connected disabilities. We’ll ask you to select all the disabilities you’re claiming for increase compensation.
2. New conditions connected to your service. 

Once you’ve listed all the conditions you are claiming, we’ll ask more specific questions about each of them.

markgreenburg commented on src/applications/disability-benefits/all-claims/config/form.js:
```diff
@@ -203,32 +218,37 @@ const formConfig = {
         disabilitiesOrientation: {
           title: '',
           path: 'disabilities/orientation',
+          // Only show the page if both (or potentially neither) options are chosen on the claim-type page
```
Is this possible though? The claimsType page uses `validateBooleanGroup` to ensure at least one selection is made (https://github.com/department-of-veterans-affairs/vets-website/pull/9470/files#diff-1ac68cd612027c8f9af3b38cad4b7633R14)

cvalarida replied:
Is the "or potentially neither" possible? No, not right now. This was basically just a safeguard against whether we decide to make that question not required either as a part of this PR (which we didn't) or in the future.

When I originally wrote this, we weren't clear on whether the question should be required, so I made it work either way. Later, when we decided to require the claim type question, I decided to keep this around in case we change our minds later.

markgreenburg replied:
Well that seems like a good idea 💯

markgreenburg commented on src/applications/disability-benefits/all-claims/config/form.js:
```diff
@@ -203,32 +218,37 @@ const formConfig = {
           path: 'disabilities/rated-disabilities',
-          depends: hasRatedDisabilities,
+          depends: formData =>
+            hasRatedDisabilities(formData) && !newOnly(formData),
```
🤔 maybe this is just me but I had to think about `newOnly` a bit. What about `newConditionsOnly` or something to more clearly indicate what's new? 🤷‍♂️

cvalarida replied:
Ah, there you go. I like `newConditionsOnly`. :+1:
I chose `newOnly` instead of `isClaimingOnlyNewConditions` because...well, y'know...Typing.

markgreenburg commented on src/applications/disability-benefits/all-claims/config/form.js:
```diff
@@ -559,14 +580,16 @@ const formConfig = {
           path: 'training-pay-waiver',
-          depends: formData => formData.hasTrainingPay,
+          depends: formData =>
+            formData.hasTrainingPay && !increaseOnly(formData),
```
🤔 so in both flows we ask veterans if they have training pay, but in the increase flow we don't follow up on it? Kind of seems like we either should ask them about it and follow up or not ask them about it at all? https://github.com/department-of-veterans-affairs/vets-website/pull/9470/files#diff-e3260aa845f2b3d5d40e8ea9cea0e6d4L192

markgreenburg replied:
Ditto with `retirementPayWaiver`, I think. There might be other places, I didn't check everything.

cvalarida replied:
Ooh, good eye! :male_detective:

@goldenmeanie I assume we should hide the pay questions at the front of the form for increase only since we're hiding the waiver pages here. Is that right?

@markgreenburg I'll look for anywhere else that may do this. Thanks!

goldenmeanie replied:
Yes, please hide for Increase-only.

markgreenburg commented on src/applications/disability-benefits/all-claims/migrations/index.js:
```diff
@@ -0,0 +1,8 @@
+// We launched at version 1 and not version 0, so the first _real_ migration is at
+//  migrations[1]
+// NOTE: This will probably just get skipped over, but it's here to be safe
+const emptyMigration = savedData => savedData;
```
🤷‍♂️

cvalarida replied:
:man_shrugging:

Commit:
Merge branch 'master' into in-form-path-choice-cv

cvalarida commented:
![image](https://user-images.githubusercontent.com/12970166/52440610-74598280-2ad3-11e9-8f7d-3d6883fae729.png)

Commit:
Update content again

Commit:
Rename to newConditionsOnly

Commit:
Hide retirementPay and trainingPay pages for increase only

Commit:
Rename function in test
